rfcs/56-better-reflect.md

```diff
@@ -0,0 +1,318 @@
+      // implementation
+    }
+    // etc.
+    fn any
+    fn any_mut
+    fn downcast
+    fn take
+    fn is
+    fn downcast_ref
+    fn downcast_mut
+    fn set
+}
+```
+
+This trait is automatically implemented by the `#[derive(Reflect)]` macro.
+
+Note that when [trait upcasting] is implemented, the `any_mut` and other
+`Any`-related methods should be moved into a `impl dyn Reflect` block,
+`as_reflect` and `as_partial_reflect` can also be dropped with trait
+upcasting.
+
+### `PartialReflect` to `Reflect` conversion
+
+We still want to be able to use the `Any` methods on our `PartialReflect`, so we
+want a way to get a new `Reflect` from a `PartialReflect`. This is only possible
+by checking if the underlying type is the one we expect OR converting into the
+underlying type described by the `PartialReflect::type_name()` method.
+
+To do so, we have two options:
+1. Provide an expected type or access a constructor stored in type registry
+   and convert with `FromReflect` from any `Reflect` to the canonical
+   representation of a `Reflect`.
+2. Add a `into_full(self: Box<Self>) -> Option<Box<dyn Reflect>>` method to
+   `PartialReflect` that returns `Some` only for canonical types.
+
+(1) Is complex and requires a lot more design, so we'll stick with (2) for this
+RFC.
+
+`into_full` will return `None` by default, but in the `#[derive(Reflect)]`
+macro will return `Some(self)`. Converting from a proxy `PartialReflect`
+to a `Reflect` is currently impossible.
+
+## User-facing explanation
+
+`PartialReflect` let you navigate any type independently of their shape
+with the `reflect_ref` and `reflect_mut` methods. `Reflect` is a special
+case of `PartialReflect` that also let you convert into a concrete type using the
+`Any` methods. To get a `Reflect` from a `PartialReflect`, you should use
+`PartialReflect::into_full`.
+
+Any type derived with `#[derive(Reflect)]` implements both `Reflect` and
+`PartialReflect`. The only types that implements `PartialReflect` but not
+`Reflect` are "proxy" types such as `DynamicStruct`, or any third party
+equivalent.
+
+## Drawbacks
+
+- It is impossible to make a proxy `PartialReflect` into a `Reflect` without
+  knowing the underlying type.
+- This splits the reflect API, that used to be neatly uniform. It will be
+  more difficult to combine various use cases of `bevy_reflect`, requiring
+  explicit conversions. (Note that however, the previous state of things would
+  result in things getting silently broken)
+
+## Unresolved questions
+
+- Should `Reflect` still be named `Reflect`, does it need to be a subtrait of
+  `PartialReflect`?
+- Is the ability to convert from a proxy `PartialReflect` to `Reflect` a
+  hard-requirement, ie: blocker? (I think only a first attempt at implementation
+  can answer this)
+
+## Future possibilities
+
+- Could we modify the `ReflectRef` and `ReflectMut` enums to enable a sort of
+  "partial" evaluation of structures by adding a `&dyn Reflect` variant?
```

Comment:
Like this?
```rust
enum ReflectRef<'a> {
  // ...
  FullReflect(&'a dyn Reflect)
}
```

It sounds like this might only work for value types. Structs and other types will need to still use their proper variant then be converted using `into_full` if needed.

Or is that the idea? If so, we could maybe have `ReflectRef::Value` hold an enum itself between partial and full reflect objects. Or maybe that's too much nesting haha

Reply:
You are right. It doesn't really make sense for `ReflectRef` and `ReflectMut`. I still had the `ReflectProxy` enum in mind. In which a `FullReflect` would make much more sense.